Better capture application requirements through driver

jdolence opened the pull request:
<!--Provide a general summary of your changes in the title above, for
example "Add AMR unit test for cell centered fields.".  Please avoid
non-descriptive titles such as "Addresses issue #8576".-->

## PR Summary

This PR is meant to allow applications to better articulate what is needed from the input deck.  This required refactoring some things, for example disassociating things like "time" from Mesh.  Introduces a mechanism for applications to specify required and desired parameters, with appropriate reporting at runtime.

This needs to wait until after #103 is merged.

<!--Please provide at least 1-2 sentences describing the pull request in
detail.  Why is this change required?  What problem does it solve?-->

<!--If it fixes an open issue, please link to the issue here.-->

Addresses #49.

## PR Checklist

<!-- Note that some of these check boxes may not apply to all pull requests -->

- [x] Code passes cpplint
- [x] New features are documented.
- [x] Adds a test for any bugs fixed. Adds tests for new features.
- [x] Code is formatted

Yurlungur commented on src/basic_types.hpp:
```diff
@@ -33,6 +35,17 @@ using Real = double;
 enum class TaskStatus { fail, complete, incomplete };
 enum class AmrTag : int { derefine = -1, same = 0, refine = 1 };
 
+struct SimTime {
+  SimTime() = default;
+  SimTime(const Real tstart, const Real tstop, const int nmax, const int ncurr,
+          const int nout)
+      : start_time(tstart), time(tstart), tlim(tstop),
+        dt(std::numeric_limits<Real>::max()), nlim(nmax), ncycle(ncurr),
+        ncycle_out(nout) {}
+  Real start_time, time, tlim, dt;
+  int nlim, ncycle, ncycle_out;
+};
+
```
I like this.

jdolence replied:
done.

Yurlungur commented on example/calculate_pi/pi.hpp:
```diff
@@ -49,7 +51,14 @@ using parthenon::Globals::nranks;
 class CalculatePi : public Driver {
  public:
-  CalculatePi(ParameterInput *pin, Mesh *pm, Outputs *pout) : Driver(pin, pm, pout) {}
+  CalculatePi(ParameterInput *pin, Mesh *pm) : Driver(pin, pm) {
+    InitializeOutputs();
+    AppInputs_t required;
+    AppInputs_t desired;
+    desired["Pi"].push_back("radius");
+    desired["graphics"].push_back("variables");
+    pin->CheckRequiredDesired(required, desired);
+  }
```
I find it a little strange that this lives in the driver, as opposed to when the package is set in `process_packages`. I dislike that this is one extra thing that needs to be overriden per application, especially when it feels like a per-package thing instead. I think this could be done at the same time as the input file is actually read. Is there a reason not to do it that way?

jdolence replied:
see my book below :wink:

jdolence commented:
OK, high level comment on the point of this PR.  Basically, the thought was that an application developer/user shouldn't have to specify things that aren't relevant for the particular application.  For example, setting tlim in a simulation where there is no notion of time.  Having to do so is just confusing.  With this PR, there are three ways requirements are enforced (and now a way to articulate parameters that are not required to be set, but the app developer suggests setting).

1. The Parthenon infrastructure has it's own demands.  Things like nx1, for example, must be specified.  This requirement was already captured by means of error checking in the Get* methods of the ParameterInput class.  That's sufficient, so I left that alone.

2. The Parthenon infrastructure demanded more than is always required.  I dealt with that in two ways.  Where it made sense (at least for now), I changed the required to optional arguments by moving Get* calls to GetOrAdd* calls, specifying a default.  In the case of time/cycle info, I pulled it out of the Mesh class entirely (which is relevant for all simulations) and stuck it into the EvolutionDriver (where time/cycle stuff has meaning).  Within that narrowed context, I still use the Get* methods to check for the required inputs.

3. Finally, an app developer may wish to require certain fields be specified (that either Parthenon does not require or that it provides a default for).  For this case, I've implemented a very simple CheckRequired(block_name, parameter_name) method in the ParameterInput class.  The intent is that this is used in the driver constructor, which seemed like a reasonable place for an app to check to make sure it's requirements are satisfied.

3.*  I also implemented a CheckDesired method that just issues a warning about missing fields.

To @Yurlungur's comment on why at the driver level instead of the package level: my thought was that packages articulate requirements via use of the Get* methods, much like Parthenon internals.  A package developer could, instead, use GetOrAdd* methods, and that would allow some apps to use the package with the optional defaults, while others could require those fields via the CheckRequired method.

Also important to note: It is certainly not a requirement that apps specify required and/or desired parameters.  This is an opt-in capability.

Yurlungur commented:
Hmm. I don't object to this opt-in approach, exactly. But I guess I don't quite see why this is needed over and above package-level parameter input. Are there examples where an application will have different requirements than an individual package? If so, I guess something like this is worthwhile. 

The refactoring that you did that decouples things like time was much needed.

jdolence commented:
> Hmm. I don't object to this opt-in approach, exactly. But I guess I don't quite see why this is needed over and above package-level parameter input. Are there examples where an application will have different requirements than an individual package? If so, I guess something like this is worthwhile.
> 
> The refactoring that you did that decouples things like time was much needed.

So (for now at least) boundary conditions is an example.  They were required (the Mesh class constructor required them), but not all applications actually need them.  Example: pi calculator.  So they are still set in the Mesh class constructor, but they have a default (they defaulted to "none" before which caused the code to die.  Now they default to reflecting.) which means they aren't required in the input file.  But an app developer might (and probably should) require them in the input if they are relevant, as they are the in the advection example.

This whole feature (the CheckRequired/CheckDesired functions) only adds like ~30 lines to the code, so it's not like there's a lot of cost here.  And I for one, plan to use them when writing physics code on top of Parthenon.

JoshuaSBrown commented on src/outputs/vtk.cpp:
```diff
@@ -242,191 +242,8 @@ void VTKOutput::WriteContainer(Mesh *pm, ParameterInput *pin, bool flag) {
-  pin->SetReal(output_params.block_name, "next_time", output_params.next_time);
-
-  return;
+void VTKOutput::WriteOutputFile(Mesh *pm, ParameterInput *pin, SimTime *tm) {
+  throw std::runtime_error(std::string(__func__) + " is not implemented");
 }
```
Are we removing support for these file formats or am I missing something?

jdolence replied:
There was some dead code here.  We can easily add this back in later if it's something we want to support.

JoshuaSBrown commented on src/parameter_input.cpp:
```diff
@@ -876,6 +901,35 @@ void ParameterInput::ForwardNextTime(Real mesh_time) {
+    missing = (GetComment(block, name) == "# Default value added at run time");
+  }
+  if (missing) {
+    std::cout << std::endl
+              << "### WARNING in CheckDesired:" << std::endl
+              << "Parameter file missing desired field <" << block << ">/" << name
+              << std::endl
+              << std::endl;
```
If we are not requiring certain attributes in the input file and are using default values, I think it would make sense to also print the default values to help the user know what is going on behind the scenes.

Yurlungur replied:
This is a good idea. Maybe should do the same for the `GetIfExists` family of functions.

jlippuner replied:
Second

jlippuner replied:
Second

jdolence replied:
good idea!  done.

JoshuaSBrown commented on src/outputs/outputs.hpp:
```diff
@@ -109,8 +109,10 @@ class OutputType {
   // following pure virtual function must be implemented in all derived classes
-  virtual void WriteOutputFile(Mesh *pm, ParameterInput *pin, bool flag) = 0;
-  virtual void WriteContainer(Mesh *pm, ParameterInput *pin, bool flag) { return; }
+  virtual void WriteOutputFile(Mesh *pm, ParameterInput *pin, SimTime *tm) = 0;
+  virtual void WriteContainer(SimTime &tm, Mesh *pm, ParameterInput *pin, bool flag) {
+    return;
+  }
```
Since you are already changing this, it might make sense to make all the arguments const, I can't think of any reason why writing the contents to a file should change them.

jdolence replied:
They shouldn't, but this isn't quite as simple as just adding the const qualifier to all the arguments.  Doing that makes the compiler barf because, for example, calling member functions in the ParameterInput object discards the const qualifier.  Let's put this off.

JoshuaSBrown commented on src/outputs/outputs.hpp:
```diff
@@ -126,7 +128,7 @@ class OutputType {
  public:
   explicit HistoryOutput(OutputParameters oparams) : OutputType(oparams) {}
-  void WriteOutputFile(Mesh *pm, ParameterInput *pin, bool flag) override;
+  void WriteOutputFile(Mesh *pm, ParameterInput *pin, SimTime *tm) override;
```
I would also recommend changing all `override` keywords to `final` for all the children of OutputType.

jdolence replied:
not here, not now, please.

jlippuner commented on src/parameter_input.hpp:
```diff
@@ -23,6 +23,7 @@
 // information on the Athena++ input file format.
 
 #include <cstddef>
+#include <map>
```
This include doesn't seem to be needed

jlippuner commented on tst/unit/test_required_desired.cpp:
```diff
@@ -0,0 +1,88 @@
+  GIVEN("A ParameterInput object already populated") {
+    ParameterInput in;
+    std::stringstream ss;
+    ss << "<block1>" << std::endl
+       << "var1 = 0   # comment" << std::endl
+       << "var2 = 1,  & # another comment" << std::endl
+       << "       2" << std::endl
+       << "<block2>" << std::endl
+       << "var3 = 3" << std::endl
+       << "# comment" << std::endl
+       << "var4 = 4" << std::endl
+       << "var_default = 5 # Default value added at run time" << std::endl;
```
Can we also test:
- Multiple lines with `&` continuation
- `&` continuation without a comment
- `&` in a comment
- `&` in a string, e.g. 
```
<comment>
problem = This is a quick & simple test
```

jdolence replied:
good catch.  forgot to deal with this case.  it is now caught and throws a runtime_error if anything but a comment follows the continuation character.

jlippuner commented on tst/unit/test_required_desired.cpp:
```diff
@@ -0,0 +1,88 @@
+
+    WHEN("We require a paramter that has been provided") {
+      THEN("Nothing should happen") {
+        REQUIRE_NOTHROW(in.CheckRequired("block1", "var1"));
+        REQUIRE_NOTHROW(in.CheckRequired("block2", "var4"));
+        REQUIRE_NOTHROW(in.CheckRequired("block1", "var2"));
```
Shouldn't we also check that there is no output in `cout_cap`?

jdolence replied:
We could.  Could do lots of other checks too.  In general, I'm not in favor of adding a test for everything imaginable.  There is a unit test in place, and we can add to it if/when we encounter issues.

jdolence commented on src/parameter_input.cpp:
```diff
@@ -876,6 +901,35 @@ void ParameterInput::ForwardNextTime(Real mesh_time) {
+void ParameterInput::CheckDesired(std::string block, std::string name) {
+  bool missing = true;
+  if (DoesParameterExist(block, name)) {
+    missing = (GetComment(block, name) == "# Default value added at run time");
```
Probably is, but this works.

jdolence commented on src/driver/driver.cpp:
```diff
@@ -12,20 +12,27 @@
-  while ((pmesh->time < pmesh->tlim) &&
-         (pmesh->nlim < 0 || pmesh->ncycle < pmesh->nlim)) {
-    if (Globals::my_rank == 0) pmesh->OutputCycleDiagnostics();
+  while ((tm.time < tm.tlim) && (tm.nlim < 0 || tm.ncycle < tm.nlim)) {
```
done.

jdolence commented on scripts/python/phdf.py:
```diff
@@ -121,11 +121,17 @@ def __init__(self, filename):
             f = h.File(filename,'r')
             # Read in the timestep attributes
-            ts = f['Timestep']
+            ts = f['Info']
```
done.

jdolence commented on example/advection/advection.hpp:
```diff
@@ -31,6 +32,7 @@ using parthenon::MultiStageBlockTaskDriver;
+using parthenon::SimTime;
 using parthenon::StateDescriptor;
 using parthenon::TaskID;
 using parthenon::TaskList;
```
done.

jdolence commented:
this one has been sitting too long.  most comments addressed.  pulling the trigger.
